dm/pkg/checker/conn_checker.go

```diff
@@ -0,0 +1,132 @@
+// limitations under the License.
+
+package checker
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/pkg/conn"
+	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+)
+
+type connAmountChecker struct {
+	targetDB *conn.BaseDB
+	stCfgs   []*config.SubTaskConfig
+
+	getConfigConn func(stCfgs []*config.SubTaskConfig) int
+	workerName    string
+}
+
+func newConnAmountChecker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig, fn func(stCfgs []*config.SubTaskConfig) int, workerName string) connAmountChecker {
+	return connAmountChecker{
+		targetDB:      targetDB,
+		stCfgs:        stCfgs,
+		getConfigConn: fn,
+		workerName:    workerName,
+	}
+}
+
+func (c *connAmountChecker) check(ctx context.Context, checkerName string) *Result {
+	result := &Result{
+		Name:  checkerName,
+		Desc:  "check if user-specified amount of connection exceeds downstream's maximum",
+		State: StateFailure,
+	}
+	baseConn, err := c.targetDB.GetBaseConn(ctx)
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	var rows *sql.Rows
+	rows, err = baseConn.QuerySQL(tcontext.NewContext(ctx, log.L()), "SHOW GLOBAL VARIABLES LIKE 'max_connections'")
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	defer rows.Close()
+	var (
+		maxConn   int
+		_variable string
+	)
+	for rows.Next() {
+		err = rows.Scan(&_variable, &maxConn)
+		if err != nil {
+			markCheckError(result, err)
+			return result
+		}
+	}
+	neededConn := c.getConfigConn(c.stCfgs)
+	if neededConn > maxConn {
+		result.Errors = append(
+			result.Errors,
+			NewError("downstream's max_connections: %d is less than the amount %s needs: %d", maxConn, c.workerName, neededConn),
+		)
+		result.Instruction = "set larger max_connections or reduce the pool size of dm"
+		result.State = StateFailure
+	} else {
+		result.State = StateSuccess
+	}
+	return result
+}
+
+type SyncerConnAmountChecker struct {
+	connAmountChecker
+}
+
+func NewSyncerConnAmountCheker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig) RealChecker {
+	return &SyncerConnAmountChecker{
+		connAmountChecker: newConnAmountChecker(targetDB, stCfgs, func(stCfgs []*config.SubTaskConfig) int {
+			syncerConn := 0
+			for _, stCfg := range stCfgs {
+				syncerConn += stCfg.SyncerConfig.WorkerCount
+			}
+			return syncerConn
+		}, "syncer"),
+	}
+}
+
+func (s *SyncerConnAmountChecker) Name() string {
+	return "syncer_conn_amount_checker"
+}
+
+func (s *SyncerConnAmountChecker) Check(ctx context.Context) *Result {
+	return s.check(ctx, s.Name())
+}
+
+type LoaderConnAmountChecker struct {
+	connAmountChecker
+}
+
+func NewLoaderConnAmountCheker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig) RealChecker {
+	return &LoaderConnAmountChecker{
+		connAmountChecker: newConnAmountChecker(targetDB, stCfgs, func(stCfgs []*config.SubTaskConfig) int {
+			loaderConn := 0
```

Comment:
loader's checkpoint db will use some connections. Does this part need to be calculated?

Reply:
I will have a look at it.

Reply:
Every unit maintains 1 db connection for checkpoint. I added it and left a comment. PTAL again😀